syntax = "proto3";

// Description of a French city.
// next ID: 14
message FrenchCity {
  // ID of the city: "Code Officiel Géographique". E.g. "31555" for Toulouse.
  string city_id = 1;

  // French name of the city.
  string name = 2;

  // ID of the département in which the city is. E.g. "69" for Rhône, "2A" for
  // Corse du Sud, or "976" for Mayotte.
  string departement_id = 3;

  // French name of the département.
  string departement_name = 4;

  // Prefix to use with departement name.
  // TODO(marielaure): Populate old users.
  string departement_prefix = 12;

  // ID of the région in which the city is. E.g. "84" for Auvergne-Rhône-Alpes.
  // This is the "new" region, after January 2016.
  string region_id = 5;

  // Name of the region (on January 1st, 2016).
  string region_name = 6;

  // A string of dash separated postcodes associated with this city, as we also have them in the
  // Algolia `cities` index. As an example for Toulouse: `31000-31100-31200-31300-31400-31500`.
  string postcodes = 7;

  // Latitude of the center of the city (approx) in northing degrees.
  float latitude = 8;

  // Longitude of the center of the city (approx) in easting degrees.
  float longitude = 9;

  // The population of the city as of the last known data.
  int32 population = 10;

  // An "urban" score: -1 for rural cities, 0 for unknown, 1 for cities in
  // urban areas between 2k and 5k, then bigger and bigger up to 8 for Paris
  // only.
  // -1: < 2000
  // 1: >=   2000 <   5000
  // 2: >=   5000 <  10000
  // 3: >=  10000 <  20000
  // 4: >=  20000 <  50000
  // 5: >=  50000 < 100000
  // 6: >= 100000 < 200000
  // 7: >= 200000 < 2M
  // 8: Paris
  // TODO(pascal): Populate old users so that we can use it more reliably.
  int32 urban_score = 11;

  // A public transportation score: score given by people living in the city in
  // https://www.ville-ideale.fr/classements.php. 0 means unknown, otherwise, value is between
  // 0.1 (worst) and 10 (greatest).
  // As of 2018-02-07, only 100 worst scores are kept. 
  // TODO(cyrille): Populate old users so that we can use it more reliably.
  float public_transportation_score = 13;
}

// Type of geographical area (order matters).
enum AreaType {
  UNKNOWN_AREA_TYPE = 0;
  CITY = 1;
  DEPARTEMENT = 2;
  REGION = 3;
  COUNTRY = 4;
  WORLD = 5;
}

// An area in France: a city, département, région or the whole country.
message Location {
  // The city of this location or one city of the area.
  FrenchCity city = 1;

  // This specifies what type of location this is. If the field `city` is set and
  // this is a type bigger than a city, then only the larger fields of the
  // FrenchCity message are relevant.
  AreaType area_type = 3;
}

// A list of cities.
message Cities {
  repeated FrenchCity cities = 1;
}

message Departement {
  string name = 1;

  string prefix = 2;
}

// A French Région.
message Region {
  // The name of the region, as should be displayed in text.
  string name = 1;

  // A prefix to put before the name whenever we wish to say "in that region".
  // For instance "en " for "Ile de France"
  string prefix = 2;
}
